dissect/jffs/jffs2.py

```diff
@@ -0,0 +1,381 @@
+from __future__ import annotations
+
+import os
+import logging
+import stat
+import zlib
+from bisect import bisect_right
+from datetime import datetime
+from functools import cache, cached_property, lru_cache
+from operator import attrgetter
+from typing import BinaryIO, Iterator, Optional, Union
+
+from dissect.util import ts
+from dissect.util.compression import lzo
+from dissect.util.stream import AlignedStream, RunlistStream
+
+from dissect.jffs.c_jffs2 import DT_MAP, JFFS2_MAGIC_NUMBERS, c_jffs2
+from dissect.jffs.exceptions import (
+    Error,
+    FileNotFoundError,
+    NotADirectoryError,
+    NotASymlinkError,
+)
+
+log = logging.getLogger(__name__)
+log.setLevel(os.getenv("DISSECT_LOG_JFFS", "CRITICAL"))
+
+
+class JFFS2:
+    """Journalling Flash File System, version 2 implementation.
+
+    Currently supports the most common compression formats (``zlib``, ``lzo``, ``none`` and ``zero``).
+    CRC checksums are currently ignored. Compression formats ``rtime``, ``rubinmips``, ``copy``
+    and ``dynrubin`` are not supported.
+
+    References:
+        - https://sourceware.org/jffs2/
+        - https://github.com/torvalds/linux/blob/master/fs/jffs2/readinode.c
+    """
+
+    def __init__(self, fh: BinaryIO):
+        self.fh = fh
+
+        self._inodes: dict[int, list[tuple[c_jffs2.jffs2_raw_inode, int]]] = {}
+        self._dirents: dict[int, dict[bytes, list[DirEntry]]] = {}
+        self._lost_found: list[list[DirEntry]] = []
+
+        if (node := c_jffs2.jffs2_unknown_node(fh)).magic not in JFFS2_MAGIC_NUMBERS:
+            raise Error(f"Unknown JFFS2 magic: {node.magic:#x}")
+
+        self.inode = lru_cache(4096)(self.inode)
+
+        self.root = self.inode(inum=0x1, type=c_jffs2.DT_DIR, parent=None)
+
+        self._scan()
+        self._garbage_collect()
+
+    def inode(self, inum: int, type: Optional[int] = None, parent: Optional[INode] = None) -> INode:
+        return INode(self, inum, type, parent)
+
+    def get(self, path: Union[str, int], node: Optional[INode] = None) -> INode:
+        if isinstance(path, int):
+            return self.inode(path)
+
+        parts = path.split("/")
```

Comment:
```suggestion
        parts = path.encode().split(b"/")
```
So we don't have to encode each individual part in the loop.